src/appleseed.python/bindblenderexport.cpp

```diff
@@ -84,17 +92,131 @@ namespace
+    // from Blender particle and dupli instances.  Storing them in a C++
+    // std::vector allows them to be processed and inserted into the appleseed scene
+    // without the overhead of Python.
+    //
+    
+    class BlTransformLibrary
+    {
+      public:
+        BlTransformLibrary() = default;
+
+        ~BlTransformLibrary() = default;
+
+        void release()
+        {
+            delete this;
+        }
+
+        int get_size() const
+        {
+            return static_cast<int>(m_xforms.size());
+        }
+
+        void add_xform_step(
+            float               time,
+            const std::string   instance_id,
+            const bpy::list&    matrix)
+        {
+            if (time == 0.0)
+                m_xforms[instance_id] = TransformSequence();
+
+            // Ignore any Blender instance that doesn't already have an associated TransformSequence
+            auto instance = m_xforms.find(instance_id);
+
+            if (instance != m_xforms.end())
+            {
+                std::vector<double> temp;
+
+                for (size_t index = 0; index < 16; ++index)
+                {
+                    bpy::extract<float> ex(matrix[index]);
+
+                    if (!ex.check())
+                    {
+                        PyErr_SetString(PyExc_TypeError, "Incompatible key type. Only floats accepted.");
+                        bpy::throw_error_already_set();
+                    }
+
+                    temp.emplace_back(static_cast<double>(ex));
+                }
+
+                instance->second.set_transform(time, Transformd(Matrix4d().from_array(temp.data())));
+            }
+        }
+
+        bool needs_assembly()
+        {
+            // Optimize all xform sequences first
+            for (auto xform = m_xforms.begin(); xform != m_xforms.end(); ++xform)
+                xform->second.optimize();
+
+            return m_xforms.size() > 1 || m_xforms.begin()->second.size() > 1;
+        }
+
+        UnalignedTransformd get_single_transform() const
+        {
+            const Transformd xform(m_xforms.begin()->second.get_earliest_transform());
+            return UnalignedTransformd(xform);
+        }
+
+        void flush_instances(
+            Assembly*           as_main_ass,
+            const char*         ass_name)
+        {
+            for (auto xform = m_xforms.begin(); xform != m_xforms.end(); ++xform)
+            {
+                auto_release_ptr<AssemblyInstance> ass_inst = AssemblyInstanceFactory().create(
+                    xform->first.c_str(),
+                    ParamArray(),
+                    ass_name);
+
+                ass_inst->transform_sequence() = xform->second;
+
+                as_main_ass->assembly_instances().insert(ass_inst);
+
+                // Store pointers to assembly instances in case updates are needed.
+                m_ass_insts[xform->first] = as_main_ass->assembly_instances().get_by_name(xform->first.c_str());
```

Comment:
You don't need to do a lookup by name, you can first obtain the pointer from `ass_inst` before inserting it into the collection of assembly instances.

Reply:
Modified.